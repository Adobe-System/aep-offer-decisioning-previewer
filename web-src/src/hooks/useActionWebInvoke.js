import { useState, useEffect } from "react";
import actionWebInvoke from "../utils";

export const useActionWebInvoke = ({
  actionName,
  headers = {},
  params = {},
}) => {
  const [data, setData] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    // invoke Action
    actionWebInvoke(actionName, headers, params)
      .then((response) => {
        setData(response);
        setIsLoading(false);
      })
      .catch((e) => {
        setData(null);
        setIsLoading(false);
        setError(e);
      });
  }, []);
  return { data, isLoading, error };
};
