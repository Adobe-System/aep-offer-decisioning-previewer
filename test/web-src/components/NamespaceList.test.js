import React from "react";
import { Provider, defaultTheme } from "@adobe/react-spectrum";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
jest.mock("../../../web-src/src/hooks/useActionWebInvoke.js");
import { useActionWebInvoke } from "../../../web-src/src/hooks/useActionWebInvoke";
import mock from "../../actions/mock";

import NamespaceList from "../../../web-src/src/components/NamespaceList";

afterEach(cleanup);
beforeEach(() => {
  jest.clearAllMocks();
});

const onSelectionChange = jest.fn(0);
const fakeProps = {
  ims: { token: "fake-token", org: "fake-org" },
  sandboxName: "fake-sandboxName",
  onSelectionChange: onSelectionChange,
};

describe("<NamespaceList> calls custom hook", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { token: "fake-token", org: "fake-org" },
      sandboxName: "fake-sandboxName",
      onSelectionChange: onSelectionChange,
    };

    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<NamespaceList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { authorization: "fake-token", "x-gw-ims-org-id": "fake-org" },
      onSelectionChange: onSelectionChange,
    };

    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<NamespaceList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
describe("<NamespaceList> on load", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("renders correctly on load", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<NamespaceList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("has a <ProgressCircle> when loading", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<NamespaceList {...fakeProps} />);
    // screen.debug();
    expect(screen.getByLabelText("Getting Identity Namespaces")).toBeDefined();
  });
});

describe("<NamespaceList> on load", () => {
  const mockResponse = {
    data: mock.data.identityPreviewReport,
    isLoading: false,
    error: null,
  };
  it("renders correctly on data", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(
      <Provider theme={defaultTheme} colorScheme={`light`}>
        <NamespaceList {...fakeProps} />
      </Provider>
    );
    expect(screen.getByText("Select Identity Namespace")).toBeInTheDocument();
    userEvent.click(screen.getByLabelText("Select Identity Namespace"));
  });
});

describe("<NamespaceList> on error", () => {
  const mockResponse = {
    data: null,
    isLoading: false,
    error: { message: "fake-error-message" },
  };
  it("renders correctly on error", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<NamespaceList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("show error message", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<NamespaceList {...fakeProps} />);
    expect(screen.getAllByText("fake-error-message")).toBeDefined();
  });
});

describe("<NamespaceList> on data", () => {
  const mockResponse = {
    data: null,
    isLoading: false,
    error: null,
  };
  it("renders correctly on no data", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<NamespaceList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("show no data message", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<NamespaceList {...fakeProps} />);
    expect(screen.getAllByText("You have no namespaces !")).toBeDefined();
  });
});
