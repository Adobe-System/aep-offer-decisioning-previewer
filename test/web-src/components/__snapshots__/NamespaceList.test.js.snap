// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NamespaceList> calls custom hook call custom hooks with global parameters correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Identity Namespaces"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="namespace-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<NamespaceList> calls custom hook call custom hooks with global parameters correctly 2`] = `
<DocumentFragment>
  <div
    aria-label="Getting Identity Namespaces"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="namespace-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<NamespaceList> on data renders correctly on no data 1`] = `
<DocumentFragment>
  <span>
    You have no namespaces !
  </span>
</DocumentFragment>
`;

exports[`<NamespaceList> on error renders correctly on error 1`] = `
<DocumentFragment>
  <span>
    fake-error-message
  </span>
</DocumentFragment>
`;

exports[`<NamespaceList> on load renders correctly on load 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Identity Namespaces"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="namespace-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
